### PR TITLE
fix(backup): post-timeout match correctness + state-gate (closes #1433)

### LINKED ISSUE #1433: [BUG] ha_backup_create TIMEOUT_OPERATION raised even when backup completed
### 💬 What Happened?

During Phase A live-tests on PR #1424 (against my deployed RBO instance), I called `ha_backup_create` twice with names `Pre_pr1424_option_b_live` and `Pre_pr1424_option_b_live_retry`. Both calls returned `success: false` with `error.code: TIMEOUT_OPERATION` and the suggestion *"Backup may still be in progress. Check Home Assistant backup status."*

Both backups had **actually completed successfully** — they were listed in Home Assistant's backup list within seconds (verified via the HA UI: the user saw both backups present, 19 and 23 minutes old respectively, with the expected names).

This is silent-failure-as-failure (the opposite direction of the typo-vs-already-deleted ambiguity addressed in PR #1424): the operation succeeds but the tool reports failure. Wrappers that branch on `success` see a failed call and may retry — producing duplicate backups — or surface a misleading error to the user.

### 📋 Copy-Paste from Conversation

**Trigger**: Two `ha_backup_create` invocations during pre-rework safety setup. Approximate response shape on both:

```json
{
  "success": false,
  "error": {
    "code": "TIMEOUT_OPERATION",
    "message": "Backup creation timed out after <max_wait_seconds> seconds",
    "context": {"backup_job_id": "...", "name": "Pre_pr1424_option_b_live"},
    "suggestions": [
      "Backup may still be in progress. Check Home Assistant backup status."
    ]
  }
}
```

**Reproduction**: any deployed instance still exposing the legacy `ha_backup_create` tool (now refactored to `ha_manage_backup(scope='snapshot', action='create')` on master — the bug also exists in the new polymorphic shape, see source location below). The HA instance was producing backups successfully but slowly enough that the tool's internal poll loop exited before observing `state="idle"` + `event_state="completed"`.

### Root cause

`src/ha_mcp/tools/backup.py` — `_poll_backup_completion` (starts at line 174). When the polling loop exits without observing `state == "idle"` AND `event_state == "completed"`, it raises `TIMEOUT_OPERATION` directly (lines 246-256):

```python
logger.warning(f"Backup did not complete within {max_wait_seconds} seconds")
raise_tool_error(
    create_error_response(
        ErrorCode.TIMEOUT_OPERATION,
        f"Backup creation timed out after {max_wait_seconds} seconds",
        context={"backup_job_id": backup_job_id, "name": name},
        suggestions=[
            "Backup may still be in progress. Check Home Assistant backup status."
        ],
    )
)
```

The tool does NOT do a final post-timeout check against the backup list to see whether a backup with the requested `name` actually exists. The polling-loop branch at lines 206-230 already has the logic to look up the completed backup by name in the backup list and construct the success response — that same lookup, repeated once after the timeout, would catch the "slow-but-completed" case.

The suggestion text *"Check Home Assistant backup status"* even acknowledges this — but pushes the verification onto the caller rather than doing it in the tool.

### Suggested fix

Before raising `TIMEOUT_OPERATION` at line 247, do one final list-backups read and check for a backup matching the requested `name`. If found, return the same success-shape response the in-loop branch produces (with a warning that completion was observed only at the post-timeout check). If not found, raise `TIMEOUT_OPERATION` with the existing message.

`_poll_backup_completion` is a private helper with a single call site at line 333 (inside `create_backup` at line 259, reached via the `ha_manage_backup` dispatcher routing `scope='snapshot', action='create'` at line 746). A single fix in this function covers both the legacy `ha_backup_create` shape exposed by older deployed instances and the current `ha_manage_backup(scope='snapshot', action='create')` polymorphic path.

### Cross-reference

Surfaced during PR #1424 Phase A live-tests (pre-rework safety setup). Out of scope for that PR; tracking here as a follow-up.

### PR DESCRIPTION
**Pull request by Patch76**
Closes #1433

## What does this PR do?

Fixes `ha_manage_backup(scope='snapshot', action='create')` reporting `TIMEOUT_OPERATION` when the backup actually completed — the wrapper-retry path then produced duplicate backups.

Scope (after KP13's review pass on the first iteration):

- **Final post-timeout `backup/info` lookup** so a backup that completed just past `max_wait_seconds` is still returned as success (with a top-level `warnings: list[str]` entry noting the late observation).
- **State-gate**: success requires `state == "idle"` AND `last_action_event.state == "completed"`. List-membership alone is insufficient — HA registers entries in `backups` before finalization, and the original fix could have claimed success on a half-written file.
- **Identity-match**: replaces name-only matching with a freshness filter (`name == name AND date >= job_start_ts - 5s`) plus authoritative-id preference via `last_action_event.backup_id`. Rejects stale prior-run entries with the same name (likely on retry); rejects concurrent UI-backups updating `last_action_event` mid-job.
- **Verification-error surfacing**: post-timeout `send_command` failures (auth, connection, command errors) now appear in the `TIMEOUT_OPERATION` context as `verification_error`, instead of being log-only.
- **Failed-state at post-timeout**: if the final lookup observes `event_state == "failed"`, raises `SERVICE_CALL_FAILED` rather than `TIMEOUT_OPERATION` with a misleading `likely_in_progress` flag.
- **`likely_in_progress` flag**: added to `TIMEOUT_OPERATION` context when the entry exists but state still indicates creation in progress — callers should back off retries rather than compound duplicates.
- **Timeout bump**: `_BACKUP_MAX_WAIT_S` 120 → 300. The 120s default underfit slow HA instances (#1433 root cause per the analysis bot's Step 3 recommendation).

New helpers:
- `_build_success_response_if_found` — single source of truth for "valid completion observed", owns state-gate + identity-match.
- `_parse_backup_date` — tz-aware ISO-8601 parsing for backup entry dates.

## Type of change

- [x] 🐛 Bug fix

## Testing

- [x] I have tested these changes with a LLM agent
- [x] All automated tests pass (`uv run pytest`)
- [x] Code follows style guidelines (`uv run ruff check`)

Test file `tests/src/unit/test_backup_poll_timeout.py` rewritten to 23 tests covering:
- Helper-level state-gate, identity-match, null-tolerance (`agents: None`, `backups: null`), date-parse edge cases.
- Post-timeout success+warnings; no-backup raises TIMEOUT_OPERATION; state-still-active raises with `likely_in_progress`; verification failures (CommandError + ConnectionError) surface via context; `success: False` falls through; name-collision picks fresh; failed-state raises SERVICE_CALL_FAILED.
- In-loop success; in-loop "idle+completed but list empty" race continues; in-loop failed-state raises.
- `create_backup` wrapper-seam: pins that the outer `except Exception` does not re-map success+warnings into INTERNAL_ERROR.

Combined with the 58 existing `test_backup_manager.py` + `test_backup_agent_lookup.py` tests, 81/81 backup unit tests pass locally. Ruff + mypy clean.

**Live bug-repro confirmed on a deployed RBO instance (v7.5.0)**: `ha_backup_create(name="Pre_Issue1433_LiveTest")` returned `success: false, error.code: TIMEOUT_OPERATION` after the 120s window while `sensor.backup_backup_manager_zustand` still showed `state="create_backup"`. The unit-test coverage of the new post-timeout branch is the substitute end-to-end evidence; the dev-addon Live-verification of the fix path was blocked by a host-network port-conflict between the stable and dev addons (both `host_network: true` on port 9583).

## Checklist

- [x] I have updated documentation if needed

## Future improvements

- End-to-end live verification of the warnings-bearing late-completion path on a deployed instance (blocked here by the addon port-conflict; would require resolving that first).
- `_parse_backup_date` could move to a shared helper module if other tools need similar ISO-8601 + naive-as-UTC parsing.

### CODE REVIEW

File: homeassistant-addon-dev/config.yaml [hunk 1]
```diff
@@ -20,7 +20,7 @@ hassio_api: true
 hassio_role: manager
 homeassistant_api: true
 host_network: true
-image: "ghcr.io/homeassistant-ai/ha-mcp-addon-dev-{arch}"
+image: "ghcr.io/patch76/ha-mcp-addon-dev-{arch}"  # TEMP: fork-image for issue-1433 live-test (revert before mark-ready)
```
[the remark this answers is not in the document]
Comment (Patch76) [reply homeassistant-addon-dev/config.yaml]: Already addressed at HEAD. The TEMP commit you reviewed (`4b8f3eb`) plus its workflow-side sibling (`4676db5`) were dropped via force-push; current HEAD `9ae2fd4` has `image: "ghcr.io/homeassistant-ai/ha-mcp-addon-dev-{arch}"` restored. Both TEMP commits existed only to build a fork-side dev addon for live-testing the fix path.

File: src/ha_mcp/tools/backup.py [hunk 2]
```diff
@@ -243,6 +271,30 @@ async def _poll_backup_completion(
...
+        if response is not None:
+            response["warnings"] = [
+                f"Backup completion observed only after the {max_wait_seconds}s "
+                "poll window — the operation succeeded but took longer than "
+                "expected. Increase max_wait_seconds if this recurs.",
+            ]
+            logger.info(
+                f"Backup found in post-timeout list lookup: {response['backup_id']}"
+            )
+            return response
```
[the remark this answers is not in the document]
Comment (Patch76) [reply src/ha_mcp/tools/backup.py]: Verified against `src/ha_mcp/client/websocket_client.py:471-510`: `send_command` raises `HomeAssistantCommandError` when `response["success"] is False`, so the prior `if final_info.get("success")` guard was unreachable on the failure path. Addressed in `9289562`: the post-timeout `send_command` is now wrapped in `try/except Exception` so any failure during this best-effort verification logs a warning and falls through to `TIMEOUT_OPERATION` — the original poll-window timeout stays the most accurate diagnosis when the post-timeout check itself can't complete.

File: tests/src/unit/test_backup_poll_timeout.py [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+        with pytest.raises(ToolError) as exc_info:
+            await _poll_backup_completion(
+                ws,
+                name="Any_Backup",
+                backup_job_id="job-1",
+                max_wait_seconds=0,
+                poll_interval=1,
+                agent_id="backup.local",
+            )
+        assert "TIMEOUT_OPERATION" in str(exc_info.value)
```
[the remark this answers is not in the document]
Comment (Patch76) [reply tests/src/unit/test_backup_poll_timeout.py]: Addressed in `9289562`. Test renamed to `test_post_timeout_final_info_call_raises_falls_through_to_timeout`; mock now uses `ws.send_command.side_effect = HomeAssistantCommandError("ws closed")` to match the actual raised-exception failure mode of `HomeAssistantWebSocketClient.send_command`.

File: src/ha_mcp/tools/backup.py [hunk 4]
```diff
@@ -171,6 +171,38 @@ async def _get_backup_password(
...
+) -> dict[str, Any] | None:
+    """Return the canonical success-response dict if a backup with ``name``
+    is present in ``info_result['result']['backups']``, else None.
+
+    Single source of truth for the success shape — called both from the
+    in-loop branch (when state=idle+completed is observed) and from the
+    post-timeout final check (#1433).
+    """
+    backups = info_result.get("result", {}).get("backups", [])
+    created_backup = next((b for b in backups if b.get("name") == name), None)
```
[the remark this answers is not in the document]
Comment (Patch76) [reply src/ha_mcp/tools/backup.py]: Adopted in `6100553`. `info_result.get("result", {}).get("backups") or []` — the suggested form lands the null-value-not-key-missing case correctly.

File: src/ha_mcp/tools/backup.py [hunk 5]
```diff
@@ -171,6 +171,38 @@ async def _get_backup_password(
...
+    if created_backup is None:
+        return None
+    return {
+        "success": True,
+        "backup_id": created_backup.get("backup_id"),
+        "backup_job_id": backup_job_id,
+        "name": name,
+        "date": created_backup.get("date"),
+        "size_bytes": created_backup.get("agents", {}).get(agent_id, {}).get("size"),
```
[the remark this answers is not in the document]
Comment (Patch76) [reply src/ha_mcp/tools/backup.py]: Adopted in `6100553`. `(created_backup.get("agents") or {}).get(agent_id, {}).get("size")` per your suggestion.

File: src/ha_mcp/tools/backup.py [hunk 6]
```diff
@@ -243,6 +271,40 @@ async def _poll_backup_completion(
...
+        f"Backup did not reach state=idle within {max_wait_seconds}s; "
+        "performing final backup-list lookup before raising timeout"
+    )
+    # send_command raises on failure rather than returning success=false; treat
+    # any exception during this best-effort verification as "couldn't verify"
+    # and fall through to the original timeout signal rather than letting an
+    # unrelated WebSocket error mask it.
+    try:
+        final_info = await ws_client.send_command("backup/info")
+    except Exception as e:
```
[the remark this answers is not in the document]
Comment (Patch76) [reply src/ha_mcp/tools/backup.py]: Adopted in `6100553`. Narrowed to `except HomeAssistantError` (the base class for `HomeAssistantCommandError` / `HomeAssistantCommandTimeout` / `HomeAssistantConnectionError`, defined at `src/ha_mcp/client/rest_client.py:37`). Programming errors now propagate per `.gemini/styleguide.md:43-49` polling-loop rule. Existing failure-mode test stays green since the mocked exception subclasses `HomeAssistantError`.

**Comment by Patch76** [comment]
## Implementation Summary

### Choices Made

- **Helper extraction**: `_build_success_response_if_found` is the single source of truth for the success-response dict. Called from both the in-loop branch (when `state=idle` + `event_state=completed` is observed) and from the new post-timeout final check. Eliminates ~15 lines of shape duplication and keeps the canonical shape in one place.
- **`warnings` surface**: post-timeout success carries a top-level `warnings: list[str]` entry per AGENTS.md tool-return-contract (degraded-success). The text names the poll window and suggests `max_wait_seconds` tuning so callers / the LLM have actionable next-steps for repeat occurrences. In-loop success returns without `warnings`.
- **Fail-through over fail-loud on the verification**: if the post-timeout `backup/info` itself raises `HomeAssistantError` (WS closed, HA error, command timeout), the original `TIMEOUT_OPERATION` is still the most accurate diagnosis — the verification is best-effort. Catching narrowly so programming errors (`AttributeError`, `TypeError`) propagate per `.gemini/styleguide.md:43-49` polling-loop rule.
- **Defensive null-guards on HA response shape**: `info_result.get("result", {}).get("backups") or []` and `(created_backup.get("agents") or {}).get(...).get("size")` handle the key-present-but-null case for both the `backups` list and the `agents` dict (Gemini sibling-parity finding).

### Problems Encountered

- **Initial `except Exception` was too broad.** First iteration would have swallowed `AttributeError` / `TypeError` from coding bugs. Narrowed to `except HomeAssistantError` after Gemini-flagged in the second review pass.
- **Initial mock shape was unreachable.** `test_post_timeout_final_info_call_fails_raises_timeout` originally mocked `send_command` returning `{"success": False, ...}`, but the actual API raises `HomeAssistantCommandError` on failure (`src/ha_mcp/client/websocket_client.py:471-510`). Test renamed to `test_post_timeout_final_info_call_raises_falls_through_to_timeout` + mock changed to `side_effect = HomeAssistantCommandError("ws closed")` after the same Gemini pass.
- **Local ruff 0.15.7 wraps differently than CI's 0.15.13** — required a separate format-only commit (`9ae2fd4`). No behavior change.
- **Live-test of fix path attempted but blocked.** The bug was live-reproduced on a deployed RBO instance (stable v7.5.0 still shipping the legacy `ha_backup_create`): the call returned `TIMEOUT_OPERATION` after 120s while `sensor.backup_backup_manager_zustand` was still in state `create_backup` — exact silent-failure-as-failure symptom the issue describes. Verification of the fix path against a fork-built dev addon was blocked by a `host_network: true` port-conflict (stable and dev both want host port 9583, host-network mode bypasses Docker port-remapping); the unit-test coverage of the post-timeout branch is the substitute evidence.

### Test Coverage

7 new unit tests in `tests/src/unit/test_backup_poll_timeout.py`:
- `TestBuildSuccessResponseIfFound` (3): helper returns success-shape on name-match, None on no-match, None on empty backup list.
- `TestPollBackupCompletionPostTimeout` (4): post-timeout finds-backup returns success+warnings, no-backup raises TIMEOUT_OPERATION, post-timeout `send_command` raising falls through to TIMEOUT_OPERATION, in-loop success path still works after the helper refactor.

Full backup unit suite (58 existing tests in `test_backup_manager.py` + `test_backup_agent_lookup.py`) still passes.

**Review by kingpanther13 — CHANGES_REQUESTED** [review]
Requesting changes — items below are all in files this PR is touching, so they belong inline per the Boy Scout rule rather than as follow-ups.

### `src/ha_mcp/tools/backup.py`

**L190 — name-only match can return the wrong backup on collision.** `next((b for b in backups if b.get("name") == name), None)` picks the first match. HA doesn't enforce unique names, and the post-timeout window is exactly when a stale collision is most likely (prior run timed out, retried, reused the same `name`). Tighten the match — prefer `backup_job_id` from `last_action_event` if available, else `date >= job_start`.

**L286–293 — late-check failure is invisible to the caller.** When the post-timeout `backup/info` raises `HomeAssistantError`, the exception goes only to the log; the user gets the original `TIMEOUT_OPERATION` with no signal that verification also failed. Thread the verification error into the error `context` at L316 (`"verification_error": repr(e)`) so it survives into the response.

**L300–309 — warning overstates certainty.** The late-check only reads `result.backups`, not `state` / `last_action_event.state`. A partial entry can appear while HA is still finalizing — the user sees "succeeded" and may immediately try to restore from a half-written file. Before returning success, also assert `state == "idle"` AND `event_state == "completed"`. If state is still active, keep the timeout error but add `"likely_in_progress": true` to its context.

**L194 — call out the `.get("agents") or {}` change in the commit message.** Real defensive improvement (handles `"agents": None`), but reads as accidental in the diff. One-line note in the commit body prevents future "why was this changed?" archaeology.

**L283 — `except HomeAssistantError` swallows `HomeAssistantAuthError`.** If the original timeout was auth, the late-check will catch the same auth error and re-raise `TIMEOUT_OPERATION`, which is misleading. Pairs with the L286–293 fix — surfacing `verification_error` into context makes the auth case self-documenting without narrowing the catch.

### `tests/src/unit/test_backup_poll_timeout.py`

**Missing — `final_info.get("success") is False` fall-through.** L295 doesn't check the success flag the way the in-loop branch does (L209). A future refactor could regress this into a silent `None`-return. Add `test_post_timeout_final_info_returns_success_false_falls_through`.

**Missing — in-loop "state=idle+completed but name not in list yet" → `continue` path (L264).** Same bug class as #1433, one tick earlier. A refactor could turn `continue` into `return` and ship a `None`-returning success. Add a two-poll test: first poll returns idle+completed with empty `backups`, second returns the entry.

**Missing — `create_backup` integration test.** All 7 tests call `_poll_backup_completion` directly. The user-facing `create_backup` (L326) wraps it in `try/except`. Add a test confirming the `warnings`-bearing dict survives the wrapper to the duplicate-retry caller seam — that's the surface the fix actually protects.

**Missing — same-name collision pin.** Pair with the L190 fix: assert that when two `backups` entries share `name`, the matcher picks the right one (job_id or date-based).

**Missing — state-ignored-at-final-check pin.** Once the L300–309 fix lands, pin the new behavior: list membership alone shouldn't return success if state is still active.

**Missing — `HomeAssistantConnectionError` case for the catch path.** L155 uses `HomeAssistantCommandError`. Adding a connection-error case locks in that the broad `HomeAssistantError` base is intentional and prevents a future narrowing regression.

**L73 — full-dict `==` is brittle.** Adding a field to the success shape (e.g., `compressed: bool`) would fail this test with no behavior change. Prefer subset assertions on the load-bearing keys.

**L120 — string substring vs. enum value.** `"TIMEOUT_OPERATION" in str(exc_info.value)` would silently break on enum rename. Use `ErrorCode.TIMEOUT_OPERATION.value` directly.

**L119 — pin `warnings: list[str]` at top level.** Current `"warnings" in result` only confirms presence. AGENTS.md mandates top-level `warnings: list[str]` shape — assert `isinstance(result["warnings"], list) and all(isinstance(w, str) for w in result["warnings"])`.

**Comment by Patch76** [comment]
@kingpanther13 — addressed all 14 items in commit `d645636` plus the underfit `_BACKUP_MAX_WAIT_S` from #1433's analysis (Step 3 root-cause fix that the first pass dropped). Pre-mark-ready adversarial self-review surfaced two more correctness issues that landed in the same commit.

## `src/ha_mcp/tools/backup.py`

**L190 — name-only match (collision on retry)** — Replaced with a freshness filter (`name == name AND date >= job_start_ts - 5s`) plus authoritative-id preference via `last_action_event.backup_id` within the fresh set (`backup.py:238-262`). The 5s tolerance covers clock skew. `job_start_ts` is captured at function entry (`backup.py:299`) and threaded into the helper. Stale prior-run entries with the same name are rejected; if HA exposes `last_action_event.backup_id`, that wins among fresh candidates; otherwise newest fresh by date.

**L286-293 — late-check failure invisible to caller** — Verification failures are now surfaced in the error context as `verification_error: repr(exc)` (`backup.py:355, 360-361, 412-413`). Any `HomeAssistantError` raised by the post-timeout `send_command` is captured, the message logs the error, and the resulting `TIMEOUT_OPERATION` response carries the verification-error string so the caller can differentiate "we polled out cleanly" from "we polled out AND verification itself failed."

**L300-309 — warning overstates certainty** — The success-response helper now gates uniformly on `state == "idle" AND last_action_event.state == "completed"` (`backup.py:227`) before considering any list-membership match. List membership alone is no longer sufficient — HA registers `backups` entries before finalization. When the post-timeout lookup finds the entry but state still indicates creation in progress, `likely_in_progress=true` is added to the `TIMEOUT_OPERATION` context (`backup.py:406-407, 414-415`) so callers back off retries rather than compounding duplicates.

**L194 — `.get("agents") or {}` defensive guard** — Preserved (`backup.py:267`); this reply documents the intent for git-blame archaeology. The change handles HA returning `agents: None` per entry; without it, `.get(agent_id, {})` on `None` would crash.

**L283 — `except HomeAssistantError` swallows `HomeAssistantAuthError`** — The catch class stays intentional (`backup.py:360`; covers `AuthError`, `ConnectionError`, `CommandError` uniformly under one "couldn't verify" path). The `verification_error` context-passthrough makes the auth case self-documenting without narrowing the catch — a token-expiry mid-poll now surfaces as `verification_error: 'HomeAssistantAuthError(...)'` in the TIMEOUT_OPERATION context instead of being log-only.

### Beyond your items — own adversarial pass

- **`target_id` matching must respect the freshness gate** (`backup.py:238-262`). Multi-job scenario: a concurrent UI-triggered backup updates `last_action_event` between our last in-loop poll and the final lookup; that `target_id` may match a stale collision from a prior run with the same name. Filter-first-then-prefer prevents this.
- **Post-timeout lookup observing `event_state == "failed"` raises `SERVICE_CALL_FAILED`** (`backup.py:398-405`), not `TIMEOUT_OPERATION` with a misleading `likely_in_progress=true` flag. The failure mode is known and unambiguous.

### Root-cause fix from #1433's analysis bot

- `_BACKUP_MAX_WAIT_S` 120 → 300 (`backup.py:49`). The first pass implemented only Step 1 (final check) and Step 2 (status reporting) of the analysis bot's recommendations; Step 3 (timeout bump) was dropped without justification. 120s underfit the slow RBO instance that triggered the bug; 300s covers the long tail.

## `tests/src/unit/test_backup_poll_timeout.py` — full rewrite, 23 tests

**Missing #1 — `final_info.get("success") is False` fall-through** (`test_post_timeout_final_info_success_false_falls_through`, `:443`).

**Missing #2 — in-loop "state=idle+completed but name not in list yet" continue** (`test_in_loop_state_idle_name_not_in_list_continues_then_finds`, `:536`). First poll returns idle+completed with empty `backups`; second poll picks up the entry. No `warnings` (in-loop success, not post-timeout).

**Missing #3 — `create_backup` wrapper integration test** (`TestCreateBackupWrapperSeam::test_create_backup_returns_warnings_dict_unchanged`, `:587`). Pins that the wrapper's outer `except Exception → exception_to_structured_error` does not re-map the success+warnings dict into `INTERNAL_ERROR`. Mocks `_poll_backup_completion` to return a warnings-bearing dict, asserts identity.

**Missing #4 — same-name collision pin** (`test_post_timeout_name_collision_picks_fresh_entry`, `:481`). Two `Pre_Test` entries: stale (2025-01-01) and fresh (2099-01-01). Matcher returns the fresh one.

**Missing #5 — state-ignored-at-final-check pin** — Three coordinated tests: `test_post_timeout_state_still_active_keeps_timeout_with_in_progress_flag` (`:378`) at the polling level, plus `test_returns_none_when_state_active_even_if_name_in_list` (`:150`) and `test_returns_none_when_event_state_not_completed` (`:170`) at the helper level.

**Missing #6 — `HomeAssistantConnectionError` variant** (`test_post_timeout_connection_error_falls_through`, `:422`). Pin that the broad `HomeAssistantError` catch covers `ConnectionError` alongside `CommandError`; prevents a future narrowing regression.

**L73 — full-dict `==` is brittle** — `test_returns_success_shape_when_name_matches` (`:114`) now uses subset assertions on load-bearing keys (`success`, `backup_id`, `name`, `backup_job_id`, `size_bytes`, `duration_seconds`). Adding a field to the success shape no longer breaks this test.

**L120 — string substring vs enum value** — All `pytest.raises(ToolError)` assertions use `ErrorCode.TIMEOUT_OPERATION.value` / `ErrorCode.SERVICE_CALL_FAILED.value` (`:375, :397, :417, :439, :457, :477, :576`). Survives enum renames.

**L119 — pin `warnings: list[str]` at top level** — `test_post_timeout_finds_backup_returns_success_with_warning` (`:339`) asserts `isinstance(result["warnings"], list)` AND `all(isinstance(w, str) for w in result["warnings"])` plus content-substring check (`:358-360`). Matches AGENTS.md tool-return contract.

### Own adversarial-pass additions

- `test_target_id_matching_stale_entry_falls_back_to_fresh` (`:184`) — `last_action_event.backup_id` matches a stale entry; freshness filter rejects the stale match; fresh entry with different `backup_id` wins.
- `test_post_timeout_failed_state_raises_service_call_failed` (`:460`) — post-timeout lookup observes `event_state == "failed"` → `SERVICE_CALL_FAILED`, not `TIMEOUT_OPERATION`. Also asserts `likely_in_progress` is NOT in the message (the flag would be actively misleading for an explicit failure).

### Helper-level coverage additions

11 tests on `_build_success_response_if_found` directly, covering: state-gate (state not idle, event_state not completed), identity-match preference, name-only date-filter fallback, all-stale rejection, null-tolerance on `agents` and `backups` fields, and the target_id-vs-stale freshness interaction.

---

### Verification

- 23 tests in this file + 58 existing in `test_backup_manager.py` + `test_backup_agent_lookup.py` → 81/81 pass.
- `ruff check` + `ruff format --check` clean on both edited files.
- `mypy` strict clean on `src/ha_mcp/tools/backup.py`.

### Diff stats

- `backup.py`: +165 / -62
- `test_backup_poll_timeout.py`: +488 / -39 (full rewrite)
- Combined: +653 / -101

### Step 5.5 adversarial-pass note

Per the toolkit Boy-Scout protocol, ran a pre-commit inline adversarial pass (sub-agent dispatch stalled on a transient classifier degradation; fell back to manual). Two correctness items surfaced beyond your review and are folded into the same commit (F1: target_id freshness gate; F7: failed-state post-timeout → SERVICE_CALL_FAILED). Documenting because they expand the scope mid-pass and you may want to spot-check.

**Review by kingpanther13 — CHANGES_REQUESTED** [review]
Spot-checks on the revision found one real issue and one minor pin gap. Both are in files this PR is already touching.

### `src/ha_mcp/tools/backup.py`

**L250 — `or last_event.get("backup_job_id")` fallback is unreachable.** `target_id = last_event.get("backup_id") or last_event.get("backup_job_id")` is compared at L252 only against `b.get("backup_id")`. Backup entries in `result.backups[]` carry `backup_id`, not `backup_job_id` (confirmed at L625's existing matcher and the `_backup_entry` test fixture). If `last_action_event` lacks `backup_id` and we fall to `backup_job_id`, we're comparing a job-id string against content-id strings — different ID spaces, no match, falls through to the newest-by-date branch. Either drop the `or` fallback, compare it against a `b.get("backup_job_id")` field (if HA actually exposes one on entries — verify first), or add a one-line comment explaining why the second alternative is intentional.

### `tests/src/unit/test_backup_poll_timeout.py`

**L370-385 (`test_post_timeout_no_backup_raises_timeout_operation`) — pin that `likely_in_progress` is absent on the clean idle+completed-no-match path.** The positive case is pinned by `test_post_timeout_state_still_active_keeps_timeout_with_in_progress_flag`; the negative side isn't. A future refactor that merges the two branches and unconditionally sets the flag would silently mislead callers without failing any test. Add `assert "likely_in_progress" not in str(exc_info.value)`.

**Comment by Patch76** [comment]
Addressed both items in `aec8130` — both in files this PR already touches.

### `src/ha_mcp/tools/backup.py`

**L250 — `or last_event.get("backup_job_id")` fallback is unreachable.**
✅ Dropped the `or` fallback: `target_id = last_event.get("backup_id")`. Verified the apples-vs-oranges claim — backup entries in `result.backups[]` carry `backup_id`, not `backup_job_id`, per the existing matcher at L625 (`any(b.get("backup_id") == backup_id for b in backups)`) and the `_backup_entry` test fixture at `tests/src/unit/test_backup_poll_timeout.py:81` (returns `{backup_id, name, date, agents}`). When `last_event.get("backup_id")` is missing, `target_id` is now None, the `authoritative` ternary at L251-255 short-circuits to None, and `match = authoritative or max(fresh, key=...)` at L256-259 falls through to the newest-by-date branch — same downstream behavior as the previous unreachable fallback, just without the misleading code path.

### `tests/src/unit/test_backup_poll_timeout.py`

**L370-385 — pin `likely_in_progress` absent on the clean idle+completed-no-match path.**
✅ Adopted. `test_post_timeout_no_backup_raises_timeout_operation` now binds `msg = str(exc_info.value)` and adds `assert "likely_in_progress" not in msg` after the existing `TIMEOUT_OPERATION` assertion (test file post-edit L380, with a comment explaining the negative-pin rationale). The positive case stays pinned by `test_post_timeout_state_still_active_keeps_timeout_with_in_progress_flag`.

Verification: 81/81 backup unit tests pass locally (test_backup_poll_timeout.py 23 + test_backup_manager.py + test_backup_agent_lookup.py). Ruff + mypy clean on the touched files.

**Review by kingpanther13 — APPROVED** [review]
